langchain/src/vectorstores/googlevertexai.ts

```diff
@@ -0,0 +1,682 @@
+  }
+
+  async buildUrl(): Promise<string> {
+    const projectId = await this.auth.getProjectId();
+    const url = `https://${this.endpoint}/${this.apiVersion}/projects/${projectId}/locations/${this.location}/indexes/${this.index}:upsertDatapoints`;
+    return url;
+  }
+
+  buildMethod(): string {
+    return "POST";
+  }
+
+  async request(
+    datapoints: IndexDatapoint[],
+    options: AsyncCallerCallOptions
+  ): Promise<UpsertDatapointResponse> {
+    const data: UpsertDatapointRequest = {
+      datapoints,
+    };
+    return this._request(data, options);
+  }
+}
+
+interface FindNeighborsConnectionParams extends GoogleVertexAIConnectionParams {
+  indexEndpoint: string;
+
+  deployedIndexId: string;
+}
+
+interface FindNeighborsRequestQuery {
+  datapoint: {
+    datapointId: string;
+    featureVector: number[];
+    restricts?: Restriction[];
+  };
+  neighborCount: number;
+}
+
+interface FindNeighborsRequest {
+  deployedIndexId: string;
+  queries: FindNeighborsRequestQuery[];
+}
+
+interface FindNeighborsResponseNeighbor {
+  datapoint: {
+    datapointId: string;
+    crowdingTag: {
+      crowdingTagAttribute: string;
+    };
+  };
+  distance: number;
+}
+
+interface FindNeighborsResponseNearestNeighbor {
+  id: string;
+  neighbors: FindNeighborsResponseNeighbor[];
+}
+
+interface FindNeighborsResponse extends GoogleVertexAIResponse {
+  data: {
+    nearestNeighbors: FindNeighborsResponseNearestNeighbor[];
+  };
+}
+
+class FindNeighborsConnection
+  extends GoogleVertexAIConnection<
+    AsyncCallerCallOptions,
+    FindNeighborsResponse
+  >
+  implements FindNeighborsConnectionParams
+{
+  indexEndpoint: string;
+
+  deployedIndexId: string;
+
+  constructor(params: FindNeighborsConnectionParams, caller: AsyncCaller) {
+    super(params, caller);
+
+    this.indexEndpoint = params.indexEndpoint;
+    this.deployedIndexId = params.deployedIndexId;
+  }
+
+  async buildUrl(): Promise<string> {
+    const projectId = await this.auth.getProjectId();
+    const url = `https://${this.endpoint}/${this.apiVersion}/projects/${projectId}/locations/${this.location}/indexEndpoints/${this.indexEndpoint}:findNeighbors`;
+    return url;
+  }
+
+  buildMethod(): string {
+    return "POST";
+  }
+
+  async request(
+    request: FindNeighborsRequest,
+    options: AsyncCallerCallOptions
+  ): Promise<FindNeighborsResponse> {
+    return this._request(request, options);
+  }
+}
+
+export interface PublicAPIEndpointInfo {
+  apiEndpoint?: string;
+
+  deployedIndexId?: string;
+}
+
+export interface MatchingEngineArgs
+  extends GoogleVertexAIConnectionParams,
+    IndexEndpointConnectionParams,
+    UpsertDatapointParams {
+  docstore: Docstore;
+
+  callerParams?: AsyncCallerParams;
+
+  callerOptions?: AsyncCallerCallOptions;
+
+  apiEndpoint?: string;
+
+  deployedIndexId?: string;
+}
+
+export class MatchingEngine extends VectorStore implements MatchingEngineArgs {
+  declare FilterType: Restriction[];
```

Comment:
nit: We should prefer `this["FilterType"]` instead of using `Restriction[]` below on the off-chance someone wants to subclass